File: py/src/axiomkit/io/fs/spec.py
import re
from collections.abc import Sequence
from dataclasses import dataclass
from enum import StrEnum
from pathlib import Path
from typing import TypeAlias

TypeCopyPatternSeq: TypeAlias = Sequence[str] | Sequence[re.Pattern[str]]


class EnumCopySymlinkStrategy(StrEnum):
    DEREFERENCE = "dereference"
    COPY_SYMLINKS = "copy_symlinks"
    SKIP_SYMLINKS = "skip_symlinks"


class EnumCopyFileConflictStrategy(StrEnum):
    SKIP = "skip"
    OVERWRITE = "overwrite"
    ERROR = "error"


class EnumCopyDirectoryConflictStrategy(StrEnum):
    SKIP = "skip"
    MERGE = "merge"
    ERROR = "error"


class EnumCopyPatternMode(StrEnum):
    GLOB = "glob"
    REGEX = "regex"
    LITERAL = "literal"


class EnumCopyDepthLimitMode(StrEnum):
    AT_MOST = "at_most"  # <=depth
    EXACT = "exact"  # =depth


@dataclass(frozen=True, slots=True)
class SpecCopyError:
    path: Path
    exception: Exception


@dataclass(frozen=True, slots=True)
class SpecCopyTreeResult:
    ok: bool
    errors: tuple[SpecCopyError, ...] = ()
    warnings: tuple[str, ...] = ()


@dataclass(frozen=True, slots=True)
class SpecCopyPatterns:
    """Compiled include/exclude pattern sets for files and directories."""

    patterns_include_files: TypeCopyPatternSeq | None
    patterns_exclude_files: TypeCopyPatternSeq | None
    patterns_include_dirs: TypeCopyPatternSeq | None
    patterns_exclude_dirs: TypeCopyPatternSeq | None

    @staticmethod
    def _ensure_sequence(value: Sequence[str] | str | None) -> Sequence[str] | None:
        return [value] if isinstance(value, str) else value

    @staticmethod
    def _compile(
        patterns: Sequence[str] | None,
        rule_pattern: EnumCopyPatternMode,
    ) -> TypeCopyPatternSeq | None:
        if not patterns:
            return None
        if rule_pattern is EnumCopyPatternMode.REGEX:
            return [re.compile(p) for p in patterns]
        return list(patterns)

    @classmethod
    def from_raw(
        cls,
        *,
        patterns_include_files: Sequence[str] | str | None,
        patterns_exclude_files: Sequence[str] | str | None,
        patterns_include_dirs: Sequence[str] | str | None,
        patterns_exclude_dirs: Sequence[str] | str | None,
        rule_pattern: EnumCopyPatternMode,
    ) -> "SpecCopyPatterns":
        """Build a compiled pattern set from raw inputs."""
        include_files = cls._ensure_sequence(patterns_include_files)
        exclude_files = cls._ensure_sequence(patterns_exclude_files)
        include_dirs = cls._ensure_sequence(patterns_include_dirs)
        exclude_dirs = cls._ensure_sequence(patterns_exclude_dirs)

        return cls(
            patterns_include_files=cls._compile(include_files, rule_pattern),
            patterns_exclude_files=cls._compile(exclude_files, rule_pattern),
            patterns_include_dirs=cls._compile(include_dirs, rule_pattern),
            patterns_exclude_dirs=cls._compile(exclude_dirs, rule_pattern),
        )

File: rs/crates/axiomkit_io_fs/src/spec.rs
//! Copy specification models and top-level error types.

use std::fmt;
use std::path::PathBuf;

////////////////////////////////////////////////////////////////////////////////
// #region EnumsInit

/// Symlink handling policy.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum EnumCopySymlinkStrategy {
    /// Follow the link and copy the target bytes/entries.
    Dereference,
    /// Create a symbolic link at destination (do not copy target bytes).
    CopySymlinks,
    /// Ignore symlink entries.
    SkipSymlinks,
}

/// Existing destination file conflict policy.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum EnumCopyFileConflictStrategy {
    /// Keep destination file and skip current source file.
    Skip,
    /// Replace destination file with source file.
    Overwrite,
    /// Record an error and skip this file.
    Error,
}

/// Existing destination directory conflict policy.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum EnumCopyDirectoryConflictStrategy {
    /// Do not descend/copy into an already existing destination directory.
    Skip,
    /// Reuse destination directory and continue copying children into it.
    Merge,
    /// Record an error when destination directory already exists.
    Error,
}

/// Pattern matching mode for include/exclude lists.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum EnumCopyPatternMode {
    /// Shell-like wildcards (`*`, `?`, character classes).
    Glob,
    /// Regular expression pattern.
    Regex,
    /// Exact string match.
    Literal,
}

/// Depth filter mode.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum EnumCopyDepthLimitMode {
    /// Include entries with depth `<= depth_limit`.
    AtMost,
    /// Include entries with depth exactly equal to `depth_limit`.
    Exact,
}

// #endregion
////////////////////////////////////////////////////////////////////////////////
// #region StructsAndErrors

/// Input options for `copy_tree`.
#[derive(Debug, Clone)]
pub struct SpecCopyOptions {
    /// Include patterns applied to file basename.
    pub patterns_include_files: Option<Vec<String>>,
    /// Exclude patterns applied to file basename.
    pub patterns_exclude_files: Option<Vec<String>>,
    /// Include patterns applied to directory basename.
    pub patterns_include_dirs: Option<Vec<String>>,
    /// Exclude patterns applied to directory basename.
    pub patterns_exclude_dirs: Option<Vec<String>>,
    /// Pattern interpretation mode.
    pub rule_pattern: EnumCopyPatternMode,
    /// Conflict behavior for destination files.
    pub rule_conflict_file: EnumCopyFileConflictStrategy,
    /// Conflict behavior for destination directories.
    pub rule_conflict_dir: EnumCopyDirectoryConflictStrategy,
    /// Symlink handling behavior.
    pub rule_symlink: EnumCopySymlinkStrategy,
    /// Optional maximum/target depth (depends on `rule_depth_limit`).
    pub depth_limit: Option<usize>,
    /// Depth evaluation mode.
    pub rule_depth_limit: EnumCopyDepthLimitMode,
    /// Maximum worker threads for file-copy stage.
    pub num_workers_max: Option<usize>,
    /// Keep relative source tree structure in destination.
    pub if_keep_tree: bool,
    /// Do not mutate filesystem; record what would happen.
    pub if_dry_run: bool,
}

impl Default for SpecCopyOptions {
    fn default() -> Self {
        Self {
            patterns_include_files: None,
            patterns_exclude_files: None,
            patterns_include_dirs: None,
            patterns_exclude_dirs: None,
            rule_pattern: EnumCopyPatternMode::Glob,
            rule_conflict_file: EnumCopyFileConflictStrategy::Skip,
            rule_conflict_dir: EnumCopyDirectoryConflictStrategy::Skip,
            rule_symlink: EnumCopySymlinkStrategy::CopySymlinks,
            depth_limit: None,
            rule_depth_limit: EnumCopyDepthLimitMode::AtMost,
            num_workers_max: None,
            if_keep_tree: true,
            if_dry_run: false,
        }
    }
}

/// One copy failure item with path + error text.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct SpecCopyError {
    /// Failed source or destination path.
    pub path: PathBuf,
    /// User-facing error text.
    pub exception: String,
}

/// "Top-level call failed" errors (input validation / setup stage).
#[derive(Debug)]
pub enum CopyTreeError {
    /// Invalid depth combination or value.
    InvalidDepthLimit(String),
    /// Invalid include/exclude pattern.
    InvalidPattern(String),
    /// Source path is not a directory.
    SourceNotDirectory(PathBuf),
    /// Source and destination overlap (`src` contains `dst` or vice versa).
    SourceDestinationOverlap {
        /// Normalized source directory.
        source: PathBuf,
        /// Normalized destination directory.
        destination: PathBuf,
    },
    /// Destination directory initialization failed.
    DestinationInitFailed {
        /// Destination path that failed initialization.
        path: PathBuf,
        /// Underlying IO error text.
        message: String,
    },
}

impl fmt::Display for CopyTreeError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self::InvalidDepthLimit(msg) => write!(f, "{msg}"),
            Self::InvalidPattern(msg) => write!(f, "{msg}"),
            Self::SourceNotDirectory(path) => {
                write!(f, "Source is not a directory: {}", path.display())
            }
            Self::SourceDestinationOverlap {
                source,
                destination,
            } => write!(
                f,
                "Source and destination directories overlap: {} <-> {}",
                source.display(),
                destination.display()
            ),
            Self::DestinationInitFailed { path, message } => {
                write!(
                    f,
                    "Failed to initialize destination {}: {message}",
                    path.display()
                )
            }
        }
    }
}

impl std::error::Error for CopyTreeError {}

// #endregion
////////////////////////////////////////////////////////////////////////////////
